go-support/shoppingcart/cmd/shoppingcart/shoppingcart.go

```diff
@@ -0,0 +1,193 @@
+	return &ShoppingCart{
+		cart:         make([]*domain.LineItem, 0),
+		EventEmitter: cloudstate.NewEmitter(), // TODO: the EventEmitter could be provided by the event sourced handler
+	}
+}
+
+// ItemAdded is a event handler function for the ItemAdded event.
+func (sc *ShoppingCart) ItemAdded(added *domain.ItemAdded) error { // TODO: enable handling for values
+	if item, _ := sc.find(added.Item.ProductId); item != nil {
+		item.Quantity += added.Item.Quantity
+	} else {
+		sc.cart = append(sc.cart, &domain.LineItem{
+			ProductId: added.Item.ProductId,
+			Name:      added.Item.Name,
+			Quantity:  added.Item.Quantity,
+		})
+	}
+	return nil
+}
+
+// ItemRemoved is a event handler function for the ItemRemoved event.
+func (sc *ShoppingCart) ItemRemoved(removed *domain.ItemRemoved) error {
+	if !sc.remove(removed.ProductId) {
+		// this should never happen
+		return errors.New("unable to remove product")
+	}
+	return nil
+}
+
+// Handle lets us handle events by ourselves.
+//
+// returns handle set to true if we have handled the event
+// and any error that happened during the handling
+func (sc *ShoppingCart) HandleEvent(event interface{}) (handled bool, err error) {
+	switch e := event.(type) {
+	case *domain.ItemAdded:
+		return true, sc.ItemAdded(e)
+	//case *domain.ItemRemoved:
+	//	*domain.ItemRemoved is handled by reflection
+	default:
+		return false, nil
+	}
+}
+
+// AddItem implements the AddItem command handling of the shopping cart service.
+func (sc *ShoppingCart) AddItem(c context.Context, li *shoppingcart.AddLineItem) (*empty.Empty, error) {
+	if li.GetQuantity() <= 0 {
+		return nil, fmt.Errorf("cannot add negative quantity of to item %s", li.GetProductId())
+	}
+	sc.Emit(&domain.ItemAdded{
+		Item: &domain.LineItem{
+			ProductId: li.ProductId,
+			Name:      li.Name,
+			Quantity:  li.Quantity,
+		}})
+	return &empty.Empty{}, nil
+}
+
+// RemoveItem implements the RemoveItem command handling of the shopping cart service.
+func (sc *ShoppingCart) RemoveItem(c context.Context, li *shoppingcart.RemoveLineItem) (*empty.Empty, error) {
+	if item, _ := sc.find(li.GetProductId()); item == nil {
+		return nil, fmt.Errorf("cannot remove item %s because it is not in the cart", li.GetProductId())
+	}
+	sc.Emit(&domain.ItemRemoved{ProductId: li.ProductId})
+	return &empty.Empty{}, nil
+}
+
+// GetCart implements the GetCart command handling of the shopping cart service.
+func (sc *ShoppingCart) GetCart(c context.Context, _ *shoppingcart.GetShoppingCart) (*shoppingcart.Cart, error) {
+	cart := &shoppingcart.Cart{}
+	for _, item := range sc.cart {
+		cart.Items = append(cart.Items, &shoppingcart.LineItem{
+			ProductId: item.ProductId,
+			Name:      item.Name,
+			Quantity:  item.Quantity,
+		})
+	}
+	return cart, nil
+}
+
+func (sc *ShoppingCart) Snapshot() (snapshot interface{}, err error) {
+	cart := domain.Cart{
+		Items: make([]*domain.LineItem, 0),
+	}
+	cart.Items = append(cart.Items, sc.cart...)
```

Comment:
Does this need to be variadic?

Reply:
how it can be _"implicitly handled by the Go runtime"_? – I think thats not possible.

[func append(slice []Type, elems ...Type) []Type](https://golang.org/pkg/builtin/#append) is variadic, but as far as I know, the type of the second argument, `Type`, has to be the element type of the first argument of append by the [language spec](https://golang.org/ref/spec#Appending_and_copying_slices). As `sc.cart` is of type `[]*domain.LineItem` I have to suffix the second argument with the ellipsis `...` to be _spread_ into multiple arguments. This way, I don't have to loop over `sc.cart`.

Reply:
That makes sense (I was thinking out loud to eliminate a possible source of flaw, but you have satisfied my potential concern, all good here) 👍

Reply:
>I was thinking out loud to eliminate a possible source of flaw

You're most welcome. Please do at any time.